statistics/handle/statscache.go

```diff
@@ -0,0 +1,126 @@
+package handle
+
+import (
+	"encoding/binary"
+	"sync"
+
+	"github.com/pingcap/tidb/statistics"
+	"github.com/pingcap/tidb/util/kvcache"
+	"github.com/pingcap/tidb/util/memory"
+	"github.com/pingcap/tidb/util/stringutil"
+)
+
+// statsCache caches Regions loaded from PD.
+type statsCache struct {
+	mu          sync.Mutex
+	cache       *kvcache.SimpleLRUCache
+	memCapacity int64
+	version     uint64
+	memTracker  *memory.Tracker // track memory usage.
+}
+type statsCacheKey int64
+
+func (key statsCacheKey) Hash() []byte {
+	var buf = make([]byte, 8)
+	binary.BigEndian.PutUint64(buf, uint64(key))
+	return buf
+}
+
+// newstatsCache returns a new statsCahce with capacity maxMemoryLimit(initial 1G)
+func newstatsCache(maxMemoryLimit int64) *statsCache {
+	// since newstatsCache controls the memory usage by itself, set the capacity of
+	// the underlying LRUCache to max to close its memory control
+	cache := kvcache.NewSimpleLRUCache(uint(maxMemoryLimit), 0.1, 0)
+	c := statsCache{
+		cache:       cache,
+		memCapacity: maxMemoryLimit,
+		memTracker:  memory.NewTracker(stringutil.StringerStr("statsCache"), -1),
+	}
+	return &c
+}
+
+// LookupUnlock get table with id without Lock.
+func (sc *statsCache) LookupUnlock(id int64) (*statistics.Table, bool) {
+	var key statsCacheKey = statsCacheKey(id)
+	value, hit := sc.cache.Get(key)
+	if !hit {
+		return nil, false
+	}
+	table := value.(*statistics.Table)
+	return table, true
+}
+
+// Lookup get table with id.
+func (sc *statsCache) Lookup(id int64) (*statistics.Table, bool) {
```

Comment:
when return false, it means the table is not cached, but do not keep the original meaning, i.e., do not exist in DB. So   the origianl palces calling this function work well ? such as https://github.com/pingcap/tidb/pull/18788/files#diff-6bf43090c5fdcde14140e8a40deeb0eaR224, https://github.com/pingcap/tidb/pull/18788/files#diff-6bf43090c5fdcde14140e8a40deeb0eaR386, and others.

Reply:
* in this implement，when insert a table and the mem usage is more than capacity , i delete a few table to decrease the mem usage ,but it did not truely delete the entire statistic data. it only delete the `hitogram.bucket` ,`hitogram.scalars` `CMSketch` data ,which comsume most of the statistic data mem usage. other meta data is insert to the cache.
* if some **query**(or **feedback** want to update the hitogram or other things) want to use the statistic data which is already droped. in this implementation, we can return the pesudo result and use the meta data to lazy load the statistic info.
* so if the previous implementation can get the table, this implementation can equivalently load the table,(this table might only with metadata but all the statistic data)
* histogram and cmsketch are not strightly used, it used by `Table.HistColl`'s interfaces, and can be checked in column.IsInvalid() and index.IsInvalid() if it is invalid, if it is not invalid we will load it peer 3s.
* pre implementation did load all the column statistic data at the bootstrap time，and lazy load them latter，so  the meta data might guarantee correctness,

Reply:
the problem exist in the case where the table actually is in the database, but it is not in the cache. When the [feedbacke](https://github.com/pingcap/tidb/blob/4a7bae64342ff4523132a71e994db484ecea2c4f/statistics/handle/update.go#L385) is invalid.

Reply:
when it is not in the cache, it will return `nil,false`